### PR TITLE
Add a Dynamic State chapter

### PR DESCRIPTION
**Pull request by sfricke-samsung**

### CODE REVIEW

File: chapters/dynamic_state.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+vkCreateGraphicsPipelines(viewportStateCreateInfo, &pipeline);
+
+vkBeginCommandBuffer();
+// Select the pipeline and draw with the state's static value
+vkCmdBindPipeline(pipeline);
+vkCmdDraw();
+vkEndCommandBuffer();
+```
+
+When the `VkPipeline` uses [dynamic state](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#pipelines-dynamic-state), some pipeline information can be emitted at creation time and instead set during recording of the command buffer. The new logical flow is:
```
Comment (SaschaWillems) [inline chapters/dynamic_state.md:26]: > information can be **emitted** at creation
I think this should say "omitted"

Comment (sfricke-samsung) [reply chapters/dynamic_state.md]: good catch

File: chapters/dynamic_state.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+```cpp
+// Using viewport state as an example
+VkViewport viewport = {0.0, 0.0, 32.0, 32.0, 0.0, 1.0};
+
+// Set value of state
+VkPipelineViewportStateCreateInfo viewportStateCreateInfo;
+viewportStateCreateInfo.pViewports = &viewport;
+// Create the pipeline with the state value set
+vkCreateGraphicsPipelines(viewportStateCreateInfo, &pipeline);
```
Comment (pdaniell-nv) [inline chapters/dynamic_state.md:17]: I guess this is abbreviated pseudo-code, but even so seeing "VkPipelineViewportStateCreateInfo" as a parameter to vkCreateGraphicsPipelines looks odd. It should be "VkGraphicsPipelineCreateInfo". Maybe just add a couple more lines:
```c++
VkGraphicsPipelineCreateInfo pipelineCreateInfo;
pipelineCreateInfo.pViewportState = &viewportStateCreateInfo;
vkCreateGraphicsPipelines(pipelineCreateInfo, &pipeline);
```

File: chapters/dynamic_state.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+When creating a graphics `VkPipeline` object the logical flow for setting state is:
+
+```cpp
+// Using viewport state as an example
+VkViewport viewport = {0.0, 0.0, 32.0, 32.0, 0.0, 1.0};
+
+// Set value of state
+VkPipelineViewportStateCreateInfo viewportStateCreateInfo;
+viewportStateCreateInfo.pViewports = &viewport;
```
Comment (pdaniell-nv) [inline chapters/dynamic_state.md:15]: For the particular example you're giving with VK_DYNAMIC_STATE_VIEWPORT I would add this to the pseudo-code:
```c++
viewportStateCreateInfo.viewportCount = 1;
```

Later it'll be needed even for the dynamic case.

File: chapters/dynamic_state.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+When the `VkPipeline` uses [dynamic state](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#pipelines-dynamic-state), some pipeline information can be omitted at creation time and instead set during recording of the command buffer. The new logical flow is:
+
+```cpp
+// Using viewport state as an example
+VkViewport viewport = {0.0, 0.0, 32.0, 32.0, 0.0, 1.0};
+VkDynamicState dynamicState = VK_DYNAMIC_STATE_VIEWPORT;
+
+// not used now
+VkPipelineViewportStateCreateInfo viewportStateCreateInfo;
+viewportStateCreateInfo.pViewports = nullptr;
```
Comment (pdaniell-nv) [inline chapters/dynamic_state.md:35]: Add this here, which might help developers accidently missing this requirement:
```c++
viewportStateCreateInfo.viewportCount = 1;
```

File: chapters/dynamic_state.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+// Set the state as being dynamic
+VkPipelineDynamicStateCreateInfo dynamicStateCreateInfo;
+dynamicStateCreateInfo.pDynamicStates = dynamicState;
```
Comment (pdaniell-nv) [inline chapters/dynamic_state.md:39]: I would change this a little to be more precise:
```
dynamicStateCreateInfo.dynamicStateCount = 1;
dynamicStateCount.pDynamicStates = &dynamicState;
```

File: chapters/dynamic_state.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+// Create the pipeline with state value not known
+vkCreateGraphicsPipelines(dynamicStateCreateInfo, &pipeline);
```
Comment (pdaniell-nv) [inline chapters/dynamic_state.md:41]: Again I find this a bit too abbreviated. Perhaps add a little more:
```c++
VkGraphicsPipelineCreateInfo pipelineCreateInfo;
pipelineCreateInfo.pViewportState = &viewportStateCreateInfo;
pipelineCreateInfo.pDynamicState = &dynamicStateCreateInfo;
vkCreateGraphicsPipelines(pipelineCreateInfo, &pipeline);
```

File: chapters/dynamic_state.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+> Note: Vulkan is a tool, so as with most things, and there is no single answer for this.
+
+Some implementations might have a performance loss using some certain `VkDynamicState` state over a static value, but dynamic states might prevent an application from having to create many permutations of pipeline objects which might be a bigger desire for the application.
+
+## What states are dynamic
+
+The full list of possible dynamic states can be found in [VkDynamicState](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VkDynamicState).
+
+The `VK_EXT_extended_dynamic_state`, `VK_EXT_extended_dynamic_state2`, and `VK_EXT_vertex_input_dynamic_state` extensions were added with the goal to support applications that need to reduce the number of pipeline state objects they compile and bind.
```
Comment (pdaniell-nv) [inline chapters/dynamic_state.md:65]: Perhaps add VK_EXT_color_write_enable to this list.

**Review by pdaniell-nv — COMMENTED** [review]
Looks okay overall.

File: chapters/dynamic_state.md (hunk 2 above)
Comment (sfricke-samsung) [reply chapters/dynamic_state.md]: Thanks, these are all great comments and applied them all, looks much better now

**Comment by sfricke-samsung** [comment]
Happy to make additional changes if people find more issues or have enhancements to add!

**Pull request merged**